src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs

```diff
@@ -3554,7 +3554,16 @@ internal static void LogMemberInvocation(string targetName, string name, object[
 
                 for (int i = 0; i < args.Length; i++)
                 {
-                    string value = args[i] is null ? "null" : args[i].ToString();
+                    object baseObj = PSObject.Base(args[i]);
+                    string value = baseObj switch
+                    {
+                        // The argument is null or AutomationNull.Value.
+                        null => "null",
+                        // The argument is a pure PSObject without a base object.
+                        PSObject => "PSObject{}",
+                        // The argument is not a PSObject, or is a PSObject with a base object.
+                        _ => baseObj.ToString()
```

Comment:
`PSObject` argument will no longer be a problem. But this doesn't prevent the user from crafting an object with an expensive `ToString` implementation.

How about change it to always return the type name, unless it's already a string?
```
string value = baseObj switch
{
    null => "null",
    PSObject => "PSObject{}",
    string => (string) baseObj,
    _ => baseObj.GetType().FullName,
}
```

Reply:
Yes, I like this the best.  We should not trust a ToString() implementation.

Reply:
It's certainly better.
But overall it all seems pointless.
My concerns:
1. If direct method calls used to be the only way to make a script fast, now it probably works even slower. For example, I used to replace Select-String with String.IndexOf() - now it doesn't make sense.
2. According to the documentation it's just a type of logging without scanning. Is it worth sacrificing performance for this if the scripts are scanned anyway?
3. If originally a method with real arguments was logged and such a log could be of some use, now only the signature is logged - I don't see any use of such a log, it would be much more useful to log the whole script at that time.

Reply:
@PaulHigin I gave it a try and it doesn't work well. By calling `ToString`, we get more useful information for objects of many types, such as all primitive types, Enum types, `BigInteger`, `IntPtr`, and probably more. For objects of those types, we get the true value by calling `ToString`, otherwise, all we have in the log is the type name.

So far, the only report we got is about the `PSObject`, which will be resolve by this change. For non-PSObject types, it may not be that big a deal unless a type's `ToString` method is intentionally made expensive to trigger the perf issue here. So, maybe we can go with `baseObj.ToString` and waiting on the feedback.

@iSazonov This logging has been proven helpful to find obfuscated malicious scripts. Maybe Paul can share more information about it.

Reply:
> @iSazonov This logging has been proven helpful to find obfuscated malicious scripts. Maybe Paul can share more information about it.

I understand the intention. But it kill performance. May this should be opt-in at least.

Reply:
@daxian-dbw  Ok, I agree that ToString() is needed for base object primitive types.  But for complex objects, ToString() usually defaults to `gettype().fullname`.  So I wonder if we special case primitive types and fall back to `gettype` for all other types?

Reply:
All enum types are not primitive, also `string` and `BigInteger` are not primitive. Those are the types I can think of off the top of my mind, and there are likely more that are interesting.
So, another option is to special case all those types that we can think of. If we get feedback from `Andrea` about other types, we can make changes and then backport to 7.3.

Reply:
@iSazonov We did the measurement and understand there is a perf overhead from this logging (but it's not "killing" the performance). However, enabling this logging is an internal requirement on Windows. Same logging has already been enabled to other built-in scripting hosts on Windows, such as CScript and WScript.